heap/[H]295.cpp

```diff
@@ -0,0 +1,181 @@
+- Space complexity: O(n)
+*/
+
+class MedianFinder {
+public:
+    multiset<int> s;
+    /** initialize your data structure here. */
+    MedianFinder() {
+        
+    }
+    
+    void addNum(int num) {
+        s.insert(num);
+    }
+    
+    double findMedian() {
+        int count = 0;
+        auto it = s.begin();
+        int target = (s.size() - 1) / 2;
+        // cout << "target = " << target << endl;
+        while (count++ < target) {
+            it++;
+        }
+        if (s.size() & 1) {
+            return *it;
+        } else {
+            int first = *it;
+            int second = *next(it);
+            return ((double) first + second) / 2;
+        }
+    }
+};
+
+/* C2: Sử dụng sorted array
+- Với cách 1, dù là lưu một dãy được sort nhưng có nhược điểm là thao tác findMedian có độ phức tạp quá lớn O(n) do không phải random access container. Ta có thể khắc phục
+bằng việc sử dụng vector để có thể random access.
+- Lưu một vector và đảm bảo vector này luôn có thứ tự tăng dần. Với thao tác add, thay vì ta add vào cuối và sort lại (độ phức tạp nlog(n)), ta có thể sử dụng binary search
+để tìm vị trí phù hợp trước, sau đó add vào vị trí đó chỉ mất độ phức tạp O(n).
+- Với thao tác findMedian, ta chỉ cần chọn các phần tử ở giữa và return.
+- Time complexity:
+  + addNum: O(n)
+  + findMedian: O(1)
+- Space complexity: O(n)
+*/
+class MedianFinder {
+public:
+    vector<int> _data;
+    /** initialize your data structure here. */
+    MedianFinder() {
+        
+    }
+    
+    void addNum(int num) {
+        auto it = upper_bound(_data.begin(), _data.end(), num);
+        _data.insert(it, num);
+        // for (int i = 0; i < _data.size(); ++i) {
+        //     cout << _data[i] << " ";
+        // }
+        // cout << endl;
+    }
+    
+    double findMedian() {
+        int n = _data.size();
+        if (n & 1) {
+            return _data[n / 2];
+        } else {
+            return ((double)_data[n/2-1] + _data[n/2]) / 2;
+        }
+    }
+};
+
+/* C3: Sử dụng heap
+- Với phương pháp 2, thao tác addNum vẫn còn độ phức tạp khá lớn, O(n). 
+- Ta có thể improve bằng cách sử dụng 2 heap, 1 min heap lưu nửa lớn hơn của dãy => top of heap là phần tử ở nằm giữa dãy khi được sort, 1 max heap lưu nửa nhỏ hơn của dãy
+=> top of heap là phần tử nằm giữa khi dãy được sort.
+- Ta thiết kế phù hợp để khi số phần tử lẻ, thì maxHeap luôn > minHeap 1 phần tử => top of maxHeap là median. Trường hợp số phần tử chẵn thì ta chia đều ra 2 heap
+và lấy trung bình của top of mỗi heap.
+- Time complexity:
+  + addNum: O(log(n))
+  + getMedian: O(1)
+- Space complexity: O(n)
+*/
+class MedianFinder {
+public:
+    priority_queue<int, vector<int>, greater<int>> minHeap;
+    priority_queue<int> maxHeap;
+    /** initialize your data structure here. */
+    MedianFinder() {
+        
+    }
+    
+    int dataSize() {
+        return maxHeap.size() + minHeap.size();
+    }
+    
+    void addNum(int num) {
+        if (dataSize() & 1) {
```

Comment:
PHần này có thể code đơn giản hơn. Nếu là tao, tao sẽ làm theo 3 bước đơn giản: 1) Cứ ném thẳng num vào minHeap. 2)  pop thằng to nhất của minHeap, ném sang maxHeap. 3) Nếu max heap size mà hơn min heap size 2 đơn vị thì pop thằng đỉnh của max heap ra, ném lại vào min heap. Ko cần phải if else nhiều. Với Java thì đơn giản thế này: 
public void addNum(int num) {
		minHeap.push(num);
		maxHeap.push(minHeap.pop());
		if (maxHeap.size() > minHeap.size() + 1) {
			minHeap.push(maxHeap.pop());
		}
	}

Reply:
Uh thanks chú, anh sửa rồi.
Expert google có khác, nói cái mà anh thấy cứ sáng cả lên :)